### ISSUE TITLE
Siblings dependencies

### ISSUE THREAD

**Issue by grean** [graphql-dotnet/graphql-dotnet#2282]
Hi,

**Use case** 
In an ObjectGraphType, i have a Field that resolve with a required Async sibling dependency, ie :
A patient wants to calculate his last weight measurement. So, in my model Patient, i have a property 
```c#
public Measurement LastMeasurement => Measurements?.OrderByDescending(m => m.Date).FirstOrDefault();
```
But it doesn´t work, because in the same time, i'm requesting the necessary Measurements with a dataloader.
Is there a way to tell to some siblings fields to wait for their resolution until all their dependencies are ready to be used ?

Is this solution is an anti pattern compared to graphql spec ?
If yes, what is the good way to handle such a use case ?

Thanks.

**Comment by Shane32** [comment]
The graphql spec has no bearing on this issue; your problem is determining the best way to load data from your database.

You have a couple options:

1. When the parent is being resolved, examine which child fields are being requested, and load the necessary data at that time.
2. Issue another call to the database for this field

A data loader can force a sibling node to execute after its siblings.  However, this would mean making the Measurements not load via a dataloader, and having the LastMeasurements field load via a data loader.  Probably not what you want to do if there are multiple parent objects returned in one call.

Besides that, you cannot control the order in which sibling fields are resolved.  With a serial execution strategy, fields are resolved in the order listed, but LastMeasurement might come listed before the Measurements nodes in the query.

**Comment by Shane32** [comment]
Another option is to put both the siblings in an object which has two fields: “items” and “last”. So the object resolves via data loader together, and then it is easy to return the items or the last item from the resolved list.

**Comment by sungam3r** [comment]
Consider changing your data model and pull down `last` property in data structure:
```graphql
{
  user {
    measurements {
      items
      last
    }
  }
}
```

**Comment by Shane32** [comment]
Thanks for the example. That’s exactly what I meant.

**Comment by sungam3r** [comment]
😄 I haven't even seen your comment.

**Comment by grean** [comment]
Ok, i see what you mean but, my case is a bit more complicated since i've some calculus that needs lastMeasurement but not only :(
The others dependencies are siblings too.. I could put all the calculus in your new object but it feels weird.
So i've decided to load the dependencies alltogether with the parent object and it works. But i've a question on how to load the depedencies. I've found two ways of doing that.
First one with AsSplitQuery macanism
```c#

        public async Task<Patient> GetPatientByIdAsync(Guid id, CancellationToken cancellationToken)
        {
            using (var scope = _serviceProvider.CreateScope())
            {
                var dbContext = scope.ServiceProvider.GetRequiredService<DietContext>();
                return await dbContext.Patients.AsSplitQuery()
                    .Include(i => i.Regimes)
                    .Include(i => i.Measurements)
                    .Include(i => i.Folders)
                    .Include(i => i.Nap)
                        .FirstAsync(i => i.Id == id, cancellationToken);
            }
        }
```
It loads correctly but the split SQL querys are run sequentially, plus, it duplicates my 1 to 1 Nap relation property on each of them (why this behavior with 1 to 1 relation ?)

Second one using multiple dbContext with Tasks, then i rebuild my patient object with dependencies like the dbContext would do it.
```c#
        public Patient GetPatientByIdAsync2(Guid id, CancellationToken cancellationToken)
        {
            var patientTask = _dbContextHelper.GetPatient(id, cancellationToken);
            var measurementTask = _dbContextHelper.GetMeasurements(id, cancellationToken);
            var folderTask = _dbContextHelper.GetFolders(id, cancellationToken);
            var regimeTask = _dbContextHelper.GetRegimes(id, cancellationToken);

            Task.WaitAll(new Task[] { patientTask, measurementTask, folderTask, regimeTask }, cancellationToken);
            var patient = patientTask.Result;

            patient.Measurements = measurementTask.Result;
            patient.Regimes = regimeTask.Result;
            patient.Folders = folderTask.Result;

            return patient;
        }
```
Class helper that allow to launch in parallel all the SQL queries
```c#
public async Task<Patient> GetPatient(Guid id, CancellationToken cancellationToken)
        {
            using (var scope = _serviceProvider.CreateScope())
            {
                var dbContext = scope.ServiceProvider.GetRequiredService<DietContext>();
                return await dbContext.Patients.Include(i => i.Nap).FirstAsync(i => i.Id == id, cancellationToken);
            }
        }

        public async Task<List<Regime>> GetRegimes(Guid id, CancellationToken cancellationToken)
        {
            using (var scope = _serviceProvider.CreateScope())
            {
                var dbContext = scope.ServiceProvider.GetRequiredService<DietContext>();
                return await dbContext.Regimes.Where(i => i.PatientId == id).ToListAsync(cancellationToken);
            }
        }
//remove others for brevity
```
But now my Root Query Field as change from FieldAsync to Field because my repository doesn't return a task anymore.
Does that change on Root Query Field type is a problem ?

What strategy do you prefer ?

Side question :
How do you do to make my helpers function being Templated ? like this 
```c#
        public async Task<T> GetResult<T>(Guid id, CancellationToken cancellationToken)
        {
            using (var scope = _serviceProvider.CreateScope())
            {
                var dbContext = scope.ServiceProvider.GetRequiredService<DietContext>();
                return await dbContext.Set<T>().Include(i => i.Nap).FirstAsync(i => i.Id == id, cancellationToken);
            }
        }
```
Using like this 
```c#
_dbContextHelper.GetResult<Patient>(id, cancellationToken);
```
But, i have an error like this 
> CS0452: Le type 'T' doit être un type référence afin d'être utilisé comme paramètre 'TEntity' dans le type ou la méthode générique 'DbContext.Set<TEntity>()'

Plus, my 1 to 1 relation is not reconize anymore due to templating, i should test for Type of Entity but i don't know the good way to do it :(
Thanks.

**Comment by Nobusuma** [comment]
> A data loader can force a sibling node to execute after its siblings. However, this would mean making the Measurements not load via a dataloader, and having the LastMeasurements field load via a data loader. Probably not what you want to do if there are multiple parent objects returned in one call.

I am having a problem were I have a field called stadistics and takes the data from a sibling, I think what you mentioned here can solve my issue.

Can you provide an example of forcing a dataloader to execute after it siblings?

**Comment by Shane32** [comment]
Going mostly from memory here so lmk if it’s not correct.

```cs
return new SimpleDataLoader<IDataLoaderResult<Class1>>(
    async token => {
        // old resolver code
    });
// note: returns IDataLoaderResult<IDataLoaderResult<Class1>>
```

**Comment by sungam3r** [comment]
@grean I hope you solved this issue. Closing due to inactivity.
